c/experimental/stf/include/cccl/c/experimental/stf/stf.h

```diff
@@ -0,0 +1,240 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#ifndef CCCL_C_EXPERIMENTAL
+#  error "C exposure is experimental and subject to change. Define CCCL_C_EXPERIMENTAL to acknowledge this notice."
+#endif // !CCCL_C_EXPERIMENTAL
+
+#include <assert.h>
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+// TODO use CCCL_C_EXTERN_C_BEGIN/CCCL_C_EXTERN_C_END
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum stf_access_mode
+{
+  STF_NONE  = 0,
+  STF_READ  = 1 << 0,
+  STF_WRITE = 1 << 1,
+  STF_RW    = STF_READ | STF_WRITE
+} stf_access_mode;
+
+struct stf_exec_place_device
+{
+  int dev_id;
+};
+
+struct stf_exec_place_host
+{
+  char dummy; /* dummy to keep it standard C which does not allow empty structs */
+};
+
+typedef enum stf_exec_place_kind
+{
+  STF_EXEC_PLACE_DEVICE,
+  STF_EXEC_PLACE_HOST
+} stf_exec_place_kind;
+
+struct stf_exec_place
+{
+  enum stf_exec_place_kind kind;
+  union
+  {
+    struct stf_exec_place_device device;
+    struct stf_exec_place_host host;
+  } u;
+};
+
+static inline struct stf_exec_place make_device_place(int dev_id)
+{
+  struct stf_exec_place p;
+  p.kind            = STF_EXEC_PLACE_DEVICE;
+  p.u.device.dev_id = dev_id;
+  return p;
+}
+
+static inline struct stf_exec_place make_host_place()
+{
+  struct stf_exec_place p;
+  p.kind         = STF_EXEC_PLACE_HOST;
+  p.u.host.dummy = 0; /* to avoid uninitialized memory warnings */
+  return p;
+}
+
+typedef struct stf_exec_place_device stf_exec_place_device;
+typedef struct stf_exec_place_host stf_exec_place_host;
+typedef union stf_exec_place_u stf_exec_place_u;
+typedef struct stf_exec_place stf_exec_place;
+
+struct stf_data_place_device
+{
+  int dev_id;
+};
+
+struct stf_data_place_host
+{
+  char dummy; /* dummy to keep it standard C which does not allow empty structs */
+};
+
+struct stf_data_place_managed
+{
+  char dummy; /* dummy to keep it standard C which does not allow empty structs */
+};
+
+struct stf_data_place_affine
+{
+  char dummy; /* dummy to keep it standard C which does not allow empty structs */
+};
+
+typedef enum stf_data_place_kind
+{
+  STF_DATA_PLACE_DEVICE,
+  STF_DATA_PLACE_HOST,
+  STF_DATA_PLACE_MANAGED,
+  STF_DATA_PLACE_AFFINE
+} stf_data_place_kind;
+
+struct stf_data_place
```

Comment:
1. Does it matter that the tags are user-visible members? Can't `stf_data_place` be a simple, opaque type? Something like
```c
typedef void* stf_data_place;
```
We are exposing `make_*` APIs anyway?

2. It'd be nice to still typedef it out even if we don't want opaque pointers for whatever reason:
```c
typedef struct (...) {
    ...
} stf_data_place;
```
so that we can just use `stf_data_place* data` instead of spelling out "struct" `struct stf_data_place* data` for function arguments.